Fixed-wing landing refactor

tstastny opened the pull request:
follows #19869 

## Describe problem solved by this pull request
Fixed-wing auto land issues:
- the logic is very convoluted / hard to understand what will happen given the many different parameters to set
- the altitude setpoints output from the flare curve are not well tracked (and likely *cannot be while using a feedback scheme with with TECS); further they are being ignored by open-loop maneuvers anyway

## Describe your solution
- landing logic simplified to commanding a constant glide slope to the ground from the approach entrance, and commanding a constant (small) sink rate `FW_LND_FL_SINK` once below flaring alt
- flare start is triggered by the maximum of:
  - a flare *time* based altitude, `FW_LND_FL_TIME` * descent rate, used to account for differing descent rates (e.g. due to wind)
  - OR triggered by the `FW_LND_FLALT`, if it is larger
![landing_logic drawio](https://user-images.githubusercontent.com/8026163/177310230-7befe340-4add-4af1-b29a-75c2c33d87ce.png)
- flare pitch limits and height rate commands are ramped in from the previous iteration's values at flare onset to avoid jumpy commands
- TECS controls all aspects of the auto landing airspeed and altitude/height rate, and is only constrained by pitch and throttle limits (lessening unintuitive open loop maneuver overrides)
- throttle is killed on flare, full stop.
- flare is the singular point of no return during landing
- lateral-directional guidance is active the full time, roll is only constrained by "near ground" limits given the airframe's geometry `FW_WING_SPAN`, `FW_WING_HEIGHT`
- lateral manual nudging of the touchdown point is configurable via `FW_LND_NUDGE`, allowing the operator to nudge (via remote) either the touchdown point itself (adjusting approach vector) or shifting the entire approach path to the left or right. this helps when GCS map or GNSS uncertainties set the aircraft on a slightly offset approach"
![landing_nudge drawio](https://user-images.githubusercontent.com/8026163/177294560-73174367-b196-424a-9b35-d00b1fc366b8.png)
- mission feasibility only checks to make sure the glide slope less than that corresponding to `FW_LND_ANG`. Only land waypoints with a previous waypoint type of `LOITER_TO_ALT` and `WAYPOINT` are accepted. (though the fw pos ctrl internal logic is a bit more agnostic to this).
- Desired max sink rate `FW_T_SINK_R_SP` is overridden, up to the physical limits set by `FW_T_SINK_MAX`, to accommodate the commanded glide slope (according to `FW_LND_ANG`)
- negative relative glide slope altitude now allowed to make sure that the altitude tracking outerloop continues the same descent rate even when the relative slope altitude is below the terrain. this is necessary because the vehicle (with current trajectory generation logic) will smoothly try to intersect the otherwise flattened altitude setpoint and throttle up just before flaring, which is undesirable. now the glide down will be unaffected and only the flare will cause the vehicle to disrupt the approach.
- Abort status unified into a bitmasked field `FW_LND_ABORT`
- Optional terrain following behavior during landing approach by setting `FW_LND_USETER=2`. Default is now `FW_LND_USETER=1`, sticking to a fixed glide slope reference altitude and only use the terrain estimate to detect when to flare. Results in more stable/steady landing approaches and resolves the problem of e.g. high trees below the vehicle's glide slope disrupting the glide. May still be disabled with `FW_LND_USETER=0`

## What else is missing (for another day, another PR)?:
- The glide slope will not be properly tracked (always a bit above the slope at steady state). This be due to the altitude outer loop used in TECS currently which has a feedforward *gain* on the glide slope, instead of feedforward command. Another issue is that the glide slope tracking is not speed independent. Neither of these issues are trivial to change at the moment due to the use of the trajectory generator and velocity smoother (which send state feedback independent height and height rate commands from the integrated reference trajectory). IMHO - a simple feedback based path following law for the altitude outer loop would do much better here, and make it simpler. The trajectory generator could still be used to set the "turn in" point (i.e. the height error boundary at which we stop commanding "full" down or up, and start blending in the actual glide slope of the path. But this would be recalculated each time based on current **ground speed** (or a filtered version of it, if necessary). The only issue this approach would not cover, is discrete setpoint changes on altitude (which the current velocity smoother does). However this can easily be smoothed in a slew-ish approach.
- Another smaller issue is the logic found here: https://github.com/PX4/PX4-Autopilot/blob/61cdd315a6be28981fd7a4b54d7b3ad85ba6f027/src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp#L1700-L1707
This has a *cusp* on the setpoint exactly at the point we want to track, which could lead to some small jittery setpoints. Using the path following approach outlined in the last bullet, this could be avoided, and still achieve the desired behavior by simply limiting the approach vector below the path to horizontal.

## Test data / coverage

**Landing**
- [x] demonstrate nudging in approach angle mode
- [x] demonstrate nudging in approach path mode
- [x] demonstrate landing
https://review.px4.io/plot_app?log=e45b9dfe-7483-4b66-80fc-2218d9bfd639
Video with nudging to avoid cornfield and land center on a 2m wide "runway": https://drive.google.com/file/d/1mzqC2vivPBiAp2-Ci-lTnpoCzb3lMtdH/view?usp=sharing

**VTOL**
- [x] check nothing breaks on VTOL ops

tstastny commented:
FYI @ryanjAA

ryanjAA commented:
🙏🏻🙏🏻

Epic!!

Is this what we were (attempting) to test or part of it?

Regarding the nudging, if in auto and we nudge how is RC override handled (com_rc_override)

tstastny commented:
@ryanjAA 

>Is this what we were (attempting) to test or part of it?

yes

>Regarding the nudging, if in auto and we nudge how is RC override handled (com_rc_override)

Oo. That's a good question. I've honestly never used that param. What would the ideal behavior be? I could see a couple options... if `COM_RC_OVERRIDE` is set:
1. Ignore nudging and override
2. Override only when any other stick axis (besides yaw, which is used for the nudging) is moved

On (2), we'd probably need to add a bit of a dead zone around the corresponding vertical component of the yaw stick (e.g. in mode 2, this is throttle), otherwise while nudging it will be super easy to still trigger the override. Otherwise.. we could also do:

3. Override only when the other stick (whatever stick does not have the yaw axis) is moved, but ignore e.g. if throttle is moved.

@MaEtUgR you also had some similar conflicts with multicopter nudging, right? Did you find alternative solutions?

ryanjAA commented:
> Regarding the nudging, if in auto and we nudge how is RC override handled (com_rc_override)
> >Oo. That's a good question. I've honestly never used that param. What would the ideal behavior be? I could see a couple options... if `COM_RC_OVERRIDE` is set:

I never use it either, but I brought it up since I know many do like it.

> 1. Ignore nudging and override

Probably not ideal
> 2. Override only when any other stick axis (besides yaw, which is used for the nudging) is moved

Can we nudge the pitch with this (to make it flare more?) If not then this is the ideal option.

> On (2), we'd probably need to add a bit of a dead zone around the corresponding vertical component of the yaw stick (e.g. in mode 2, this is throttle), otherwise while nudging it will be super easy to still trigger the override. Otherwise.. we could also do:

This makes the most sense and good idea about the dead zone - definitely would get triggered if I were doing it without a DZ...

> 3. Override only when the other stick (whatever stick does not have the yaw axis) is moved, but ignore e.g. if throttle is moved.

This is an OK option as well but I'm a bit torn TBH - intuitively speaking if I want to abort, I will prob throttle up, so disregarding throttle seems suboptimal. On the other hand, I would need to ensure to not be dumb with the throttle while nudging to not kick it into position mode (rc over-ride) but since both of these require steps to commit to memory, I think it's just something that needs to be learnt and abided by in either case. 
A 30% deadline on throttle is probably the way to go, you really don't move throttle 30% unintentionally....

> @MaEtUgR you also had some similar conflicts with multicopter nudging, right? Did you find alternative solutions?


Excited about this!

sfuhrer commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -1027,3 +974,59 @@ PARAM_DEFINE_FLOAT(FW_WING_SPAN, 6.0);
+/**
+ * Landing flare time
+ *
+ * Multiplied by the descent rate to calculate a dynamic altitude at which
```
I would add a comment here that only the max() of FW_LND_FLALT and FW_LND_FL_TIME* vz is taken.

tstastny replied:
done [dec33ca](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/dec33ca69ea7f98093d0c3880343330849d58560)

sfuhrer commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -448,35 +424,6 @@ PARAM_DEFINE_FLOAT(FW_LND_HVIRT, 10.0f);
  */
 PARAM_DEFINE_FLOAT(FW_LND_FLALT, 3.0f);
```
3 seems a bit high. Now that we also have the time*vz based threshold I would reduce that one and more rely on the time based one. 
E.g. 2?

sfuhrer replied:
And add also a comment here about the max() of the two is taken.

tstastny replied:
reduced, documented [dec33ca](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/dec33ca69ea7f98093d0c3880343330849d58560)

sfuhrer commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -1673,337 +1664,187 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
+
+		// open the desired max sink rate to encompass the glide slope if within the aircraft's performance limits
+		const float glide_slope_sink_rate = airspeed_land * sinf(math::radians(_param_fw_lnd_ang.get()));
+		const float desired_max_sinkrate = math::min(math::max(glide_slope_sink_rate, _param_sinkrate_target.get()),
```
Do we even need to take _param_sinkrate_target into account here?

sfuhrer commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -1673,337 +1664,187 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
+		/* longitudinal guidance */
+
+		// open the desired max sink rate to encompass the glide slope if within the aircraft's performance limits
+		const float glide_slope_sink_rate = airspeed_land * sinf(math::radians(_param_fw_lnd_ang.get()));
```
Isn't the glide slop angle given through geometry (distance loiter to alt to land point) and only constrained by _param_fw_lnd_ang?

tstastny replied:
thanks for catching this! was an oversight during a refactor [413e875](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/413e87536f0411ed1575059779261e201e06669a)

sfuhrer commented on src/modules/navigator/mission_feasibility_checker.cpp:
```diff
@@ -499,59 +510,87 @@ MissionFeasibilityChecker::checkFixedWingLanding(const mission_s &mission, bool
+						mavlink_log_critical(_navigator->get_mavlink_log_pub(),
+								     "Reduce the glide slope, lower the entrance altitude %d meters, or increase the landing approach distance %d meters.\t",
+								     (int)acceptable_entrance_alt, (int)acceptable_landing_dist);
+						events::send<uint32_t, uint32_t>(events::ID("navigator_mis_correct_glide_slope"), {events::Log::Error, events::LogInternal::Info},
```
This second info message with the recommendations is super useful actually! :rocket:

sfuhrer reviewed (commented):
Played around on SITL, feels quite intuitive! Agree though that we need to look into TECS to make it track the slope and not just the height rate.
Only found one "bug", where without NPFG the "far in the distance" landing point projection is not correct.

tstastny commented on msg/position_controller_landing_status.msg:
```diff
@@ -1,9 +1,3 @@
-float32 flare_length
-
-bool abort_landing				# true if landing should be aborted
+uint64 timestamp # [us] time since system start
```
Yes only for logging at the moment. I think that's still important though. I also wanted to add more context to the abort landing bool -- maybe an enum with different abort reasons later once we can integrate more automated aborts (something i wanted to add in this pr.. but ran out of time).

sfuhrer replied:
okay then let's keep it.

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -1673,337 +1664,187 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
+		_att_sp.fw_control_yaw = true;
 
-	} else {
+		_att_sp.thrust_body[0] = 0.0f;
```
done [6ebb20a](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/6ebb20a41f1ee7a6508a1f822c7e0feaf12516d0)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -1673,337 +1664,187 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
+		// enable direct yaw control using rudder/wheel
+		_att_sp.fw_control_yaw = false;
+
+		_att_sp.thrust_body[0] = (_landed) ? 0.0f : get_tecs_thrust();
```
done [6ebb20a](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/6ebb20a41f1ee7a6508a1f822c7e0feaf12516d0)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -2694,6 +2538,154 @@ FixedwingPositionControl::calculateTakeoffBearingVector(const Vector2f &launch_p
+			local_approach_entrance = local_position;
+		}
+
+		if (height_above_land_point < FLT_EPSILON) {
```
done [fde6bbf](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/fde6bbf5298042c84fbd33f1a2b523eb757c3389)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -2694,6 +2538,154 @@ FixedwingPositionControl::calculateTakeoffBearingVector(const Vector2f &launch_p
+
+	// if _param_fw_lnd_nudge.get() == 0, no nudging
+
+	if (_param_fw_lnd_nudge.get() == 1) {
```
done [1f401e8](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/1f401e896af0c8872702d54b534eed20c8857f88)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -2694,6 +2538,154 @@ FixedwingPositionControl::calculateTakeoffBearingVector(const Vector2f &launch_p
+{
+	float terrain_alt = land_point_altitude;
+
+	if (_param_fw_lnd_useter.get() == 1) {
```
see [927412d](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/927412d4940330dd99a1400dd2c45ef430ef5510)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp:
```diff
@@ -121,6 +120,9 @@ static constexpr float HDG_HOLD_MAN_INPUT_THRESH = 0.01f;
 // [us] time after which we abort landing if terrain estimate is not valid
 static constexpr hrt_abstime TERRAIN_ALT_TIMEOUT = 1_s;
 
+// [us] if no altimeter measurement is made within this timeout, the land waypoint altitude is used for terrain altitude
```
clarified [927412d](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/927412d4940330dd99a1400dd2c45ef430ef5510)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -2694,6 +2538,154 @@ FixedwingPositionControl::calculateTakeoffBearingVector(const Vector2f &launch_p
+			} else {
+				// still no valid terrain, abort landing
+				terrain_alt = land_point_altitude;
+				abort_landing(true);
```
refactored [927412d](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/927412d4940330dd99a1400dd2c45ef430ef5510)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp:
```diff
@@ -291,34 +303,29 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
-	bool _land_motor_lim{false};
-	bool _land_onslope{false};
-	bool _land_abort{false};
+	hrt_abstime _time_started_landing{0}; // [us]
```
done [927412d](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/927412d4940330dd99a1400dd2c45ef430ef5510)

tstastny commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -1027,3 +974,59 @@ PARAM_DEFINE_FLOAT(FW_WING_SPAN, 6.0);
+ *
+ * Approach angle nudging: shifts the touchdown point laterally while keeping the approach entrance point constant
+ * Approach path nudging: shifts the touchdown point laterally along with the entire approach path
+ *
```
added [1f401e8](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/1f401e896af0c8872702d54b534eed20c8857f88)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -1673,337 +1664,187 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
-				// normal navigation
-				_l1_control.navigate_waypoints(prev_wp_local, curr_wp_local, curr_pos_local, ground_speed);
-			}
+			const float along_track_distance_from_entrance = local_approach_entrance.unit_or_zero().dot(
```
thanks for catching this! another missed one during the last refactor [ec085af](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/ec085af3830e8c00688fe23e192f1afcf0c5b8ff)

tstastny commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp:
```diff
@@ -1673,337 +1664,187 @@ FixedwingPositionControl::control_auto_takeoff(const hrt_abstime &now, const flo
+						local_position - local_approach_entrance);
+
+			const Vector2f virtual_waypoint = local_approach_entrance + (along_track_distance_from_entrance + HDG_HOLD_DIST_NEXT) *
+							  local_approach_entrance.unit_or_zero();
```
[ec085af](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/ec085af3830e8c00688fe23e192f1afcf0c5b8ff)

sfuhrer commented on src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp:
```diff
@@ -117,10 +117,12 @@ static constexpr float HDG_HOLD_YAWRATE_THRESH = 0.15f;
 static constexpr hrt_abstime TERRAIN_ALT_TIMEOUT = 1_s;
 
-// [us] if no altimeter measurement is made within this timeout, the land waypoint altitude is used for terrain altitude
+// [us] within this timeout, if a distance sensor measurement not yet made, the land waypoint altitude is used for terrain
```
We should mention in the docs at least that the slope shouldn't start above the max distance sensor range.

tstastny replied:
yeah, agreed. at least until we get a next PR going with utilization of the driver specs for max/min range

sfuhrer replied:
Right. Let's also add a comment in the param description of FW_LND_USETER.

tstastny replied:
done in [9161270](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/91612704277675ae59f91d370b34642a9d8f682b)

tstastny commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -400,14 +400,15 @@ PARAM_DEFINE_FLOAT(FW_TKO_PITCH_MIN, 10.0f);
 /**
  * Landing flare altitude (relative to landing altitude)
  *
+ * NOTE: max(FW_LND_FLALT, FW_LND_FL_TIME * |z-velocity|) is taken as the flare altitude
```
ahh oops you're right it happens to work out with the "down" definition of z-vel. will change

tstastny commented on msg/position_controller_landing_status.msg:
```diff
@@ -1,3 +1,11 @@
+
+# abort reasons
+# corresponds to individual bits of param FW_LND_ABORT
+uint8 kAbortReasonNone = 0
```
reworked in [bcc99ea](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/bcc99eab2f6f5e9ca9ade2ec9a5eae17544a6884)

tstastny commented:
Incorporated some last additions from our field tests last week and added flight log + video in description. This would be ready to go from my side -- @dagar @sfuhrer you may want to check if you like the abort status handling i refactored.. tried to clean it up a bit and make it more descriptive.

@ryanjAA @dagar - another nice addition @sfuhrer and I thought of somehow only on the field:
>Optional terrain following behavior during landing approach via FW_LND_TER_REL. Default is now to stick to a fixed glide slope reference altitude and only use the terrain estimate to detect when to flare. Results in more stable/steady landing approaches and resolves the problem of e.g. high trees below the vehicle's glide slope disrupting the glide.

We were discussing this problem of "what to do about trees and not so nice terrain profiles during landing approach" ... the solution seems to be quite simple.. just only use the terrain estimate for detecting the flare altitude :D

sfuhrer reviewed (dismissed):
Only minor comments (I would by default enable the distance sensor for flaring). SITL tested VTOL again without noticing any regression. 
Good to got in (together with https://github.com/PX4/PX4-Autopilot/pull/19869)

tstastny commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -1068,3 +1068,17 @@ PARAM_DEFINE_INT32(FW_LND_NUDGE, 0);
+ * @boolean
+ * @group FW L1 Control
+ */
+PARAM_DEFINE_INT32(FW_LND_TER_REL, 0);
```
Yeah I wouldn't really want to use it myself. Just didnt want to kill off functionality if someone was actively using it. @dagar should we keep or delete?

tstastny replied:
alright, absorbed into `FW_LND_USETER` [3b0f141](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/3b0f141216c406b891b21bec6e65b4bdfbf56db3)

tstastny commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -1068,3 +1068,17 @@ PARAM_DEFINE_INT32(FW_LND_NUDGE, 0);
  * @group FW L1 Control
  */
 PARAM_DEFINE_INT32(FW_LND_ABORT, 0);
+
```
I purposely did not put all of that logic into the FW_LND_USETER description, because all that param does is say use the distance sensor - it does not specify what abort logic is used, or what FW_LND_TER_REL does. Those are independent. Could maybe just note there that if USETER = 0, then no terrain based abort logic is used, and no terrain following on glide slope is possible.

tstastny replied:
>only used for flaring start if FW_LND_TER_REL ==1

Just fyi - It is opposite. FW_LND_TER_REL==1 means that we listen to the terrain estimate the entire approach. Which is why it is defaulted to 0 (so that we only use it for flare trigger).

>And I would make FW_LND_USETER 1 by default.

Just to confirm - this means that folks without distance sensors (putting all default settings) will still land without aborting at the land waypoint altitude (since the default FW_LND_ABORT bits are none). Maybe we want default FW_LND_ABORT bits to be all active if the distance sensor is enable? Then it will abort if no distance sensor exists, and the user needs to actively disable the unsafe things?

sfuhrer replied:
I would though mention about flaring in the description of FW_LND_USETER, as this is the main reason to have the distance sensor enabled.

sfuhrer replied:
And yes, I meant FW_LND_TER_REL==0 :+1:

sfuhrer replied:
great!

tstastny commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -1045,9 +1045,9 @@ PARAM_DEFINE_FLOAT(FW_LND_TD_OFF, 3.0);
  * @value 1 Nudge approach angle
  * @value 2 Nudge approach path
- * @group FW L1 Control
+ * @group FW L1 ControlW
```
yup :S

tstastny replied:
resolved in [1e269d6](https://github.com/PX4/PX4-Autopilot/pull/19871/commits/1e269d63c55d298d92198b838f973ec9d19d2aa3)
